plugins/inputs/nsdp/nsdp.go

```diff
@@ -0,0 +1,121 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package nsdp
+
+import (
+	_ "embed"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/tdrn-org/go-nsdp"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+const pluginName = "nsdp"
+
+const defaultTimeout = config.Duration(2 * time.Second)
+
+type NSDP struct {
+	Target      string          `toml:"target"`
+	DeviceLimit uint            `toml:"device_limit"`
+	Timeout     config.Duration `toml:"timeout"`
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func defaultNSDP() *NSDP {
+	return &NSDP{
+		Target:  nsdp.IPv4BroadcastTarget,
+		Timeout: defaultTimeout,
+	}
+}
+
+func (*NSDP) SampleConfig() string {
+	return sampleConfig
+}
+
+func (plugin *NSDP) Init() error {
+	return nil
+}
+
+func (plugin *NSDP) Gather(acc telegraf.Accumulator) error {
+	plugin.Log.Debugf("Querying %s", plugin.Target)
+	conn, request, err := plugin.prepareGatherRequest()
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	responses, err := conn.SendReceiveMessage(request)
+	if err != nil {
+		return err
+	}
+	for device, response := range responses {
+		plugin.Log.Debugf("Processing device: %s", device)
+		plugin.gatherDevice(acc, device, response)
+	}
+	return nil
+}
+
+func (plugin *NSDP) prepareGatherRequest() (*nsdp.Conn, *nsdp.Message, error) {
+	conn, err := nsdp.NewConn(plugin.Target, plugin.Log.Level().Includes(telegraf.Debug))
+	if err != nil {
+		return nil, nil, err
+	}
+	conn.ReceiveDeviceLimit = plugin.DeviceLimit
+	conn.ReceiveTimeout = time.Duration(plugin.Timeout)
+	request := nsdp.NewMessage(nsdp.ReadRequest)
+	request.AppendTLV(nsdp.EmptyDeviceModel())
+	request.AppendTLV(nsdp.EmptyDeviceName())
+	request.AppendTLV(nsdp.EmptyDeviceIP())
+	request.AppendTLV(nsdp.EmptyPortStatistic())
+	return conn, request, nil
+}
```

Comment:
Could you reuse the connection and requests or does this really need to be done in every gather cycle?

Reply:
Adapted via 8c321f127f426a5c1cbdc9876b56cf4409fafce2
Connection is now cached and re-established on error.